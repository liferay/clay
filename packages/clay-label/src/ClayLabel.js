import 'clay-button';
import 'clay-link';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayLabel.soy.js';

/**
 * Metal Clay Label component.
 * @extends Component
 */
class ClayLabel extends Component {
	/**
	 * Handle `click` button and emit event `close`.
	 * @protected
	 */
	_handleCloseButtonClick() {
		this.emit('close');
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayLabel.STATE = {
	/**
	 * True or false to activate the close button.
	 * @default false
	 * @instance
	 * @memberof ClayLabel
	 * @type {?bool}
	 */
	closeable: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabel
	 * @type {?(string|undefined)}
	 */
	elementClasses: Config.string(),

	/**
	 * HREF in label.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabel
	 * @type {?(string|undefined)}
	 */
	href: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabel
	 * @type {?(string|undefined)}
	 */
	id: Config.string(),

	/**
	 * The label of the badge element.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabel
	 * @type {?(html|string|undefined)}
	 */
	label: Config.any(),

	/**
	 * The size of the label element.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabel
	 * @type {?(string|undefined)}
	 */
	size: Config.oneOf(['lg']),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @default undefined
	 * @instance
	 * @memberof ClayLabel
	 * @type {?(string|undefined)}
	 */
	spritemap: Config.string(),

	/**
	 * Label style color.
	 * @default secondary
	 * @instance
	 * @memberof ClayLabel
	 * @type {?string}
	 */
	style: Config.oneOf([
		'danger',
		'info',
		'secondary',
		'success',
		'warning',
	]).value('secondary'),
};

defineWebComponent('clay-label', ClayLabel);

Soy.register(ClayLabel, templates);

export {ClayLabel};
export default ClayLabel;
