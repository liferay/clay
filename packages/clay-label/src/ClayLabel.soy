{namespace ClayLabel}

/**
 * This renders a label.
 */
{template .render}
	{@param? _handleCloseButtonClick: any}
	{@param? closeable: bool}
	{@param? elementClasses: string}
	{@param? href: string}
	{@param? id: string}
	{@param? label: html|string}
	{@param? size: string}
	{@param? spritemap: string}
	{@param? style: string}

	{let $attributes kind="attributes"}
		class="label
			{sp}label-{$style ?: 'secondary'}

			{if $closeable}
				{sp}label-dismissible
			{/if}

			{if $elementClasses}
				{sp}{$elementClasses}
			{/if}

			{if $size}
				{sp}label-{$size}
			{/if}
		"

		{if $id}
			id="{$id}"
		{/if}
	{/let}

	<span {$attributes}>
		{call .content}
			{param _handleCloseButtonClick: $_handleCloseButtonClick /}
			{param closeable: $closeable /}
			{param href: $href /}
			{param label: $label /}
			{param spritemap: $spritemap /}
		{/call}
	</span>
{/template}

/**
 * Render icon and label content.
 */
{template .content}
	{@param? _handleCloseButtonClick: any}
	{@param? closeable: bool}
	{@param? href: string}
	{@param? label: html|string}
	{@param? spritemap: string}

	{let $labelContent kind="html"}
		<span class="label-item label-item-expand">
			{if $href}
				{call ClayLink.render}
					{param href: $href /}
					{param label: $label /}
				{/call}
			{else}
				{$label}
			{/if}
		</span>
	{/let}

	{if $closeable and $spritemap}
		{$labelContent}

		<span class="label-item label-item-after">
			{call ClayButton.render}
				{param ariaLabel kind="text"}
					{msg desc="The word 'Close' used as a verb"}
						close
					{/msg}
				{/param}

				{param elementClasses: 'close' /}
				{param events: [ 'click': $_handleCloseButtonClick] /}
				{param icon: 'times' /}
				{param iconAlignment: 'right' /}
				{param ref: 'closeButton' /}
				{param spritemap: $spritemap /}
				{param style: 'unstyled' /}
				{param type: 'button' /}
			{/call}
		</span>
	{else}
		{$labelContent}
	{/if}
{/template}
