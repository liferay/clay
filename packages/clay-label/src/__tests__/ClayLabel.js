'use strict';

import ClayLabel from '../ClayLabel';

let label;

describe('ClayLabel', function() {
	afterEach(() => {
		if (label) {
			label.dispose();
		}
	});

	it('should render the default markup', () => {
		label = new ClayLabel();

		expect(label).toMatchSnapshot();
	});

	it('should render a label with classes', () => {
		label = new ClayLabel({
			elementClasses: 'my-custom-class',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a label with id', () => {
		label = new ClayLabel({
			id: 'myId',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a label with label', () => {
		label = new ClayLabel({
			label: 'Foo',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a large label', () => {
		label = new ClayLabel({
			size: 'lg',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a label with removable', () => {
		label = new ClayLabel({
			label: 'Foo',
			closeable: true,
			spritemap: 'icons.svg',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a label with link', () => {
		label = new ClayLabel({
			label: 'Foo',
			href: '#1',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a `default color` label', () => {
		label = new ClayLabel({
			style: 'secondary',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a `info color` label', () => {
		label = new ClayLabel({
			style: 'info',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a `warning color` label', () => {
		label = new ClayLabel({
			style: 'warning',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a `danger color` label', () => {
		label = new ClayLabel({
			style: 'danger',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a `success color` label', () => {
		label = new ClayLabel({
			style: 'success',
		});

		expect(label).toMatchSnapshot();
	});

	it('should render a closeable label and emit an event on close button click', () => {
		label = new ClayLabel({
			label: 'Foo',
			closeable: true,
			spritemap: 'icons.svg',
		});

		const spy = jest.spyOn(label, 'emit');

		label.refs.closeButton.element.click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith('close');
	});
});
