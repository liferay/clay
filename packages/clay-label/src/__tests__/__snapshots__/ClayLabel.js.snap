// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayLabel should render a \`danger color\` label 1`] = `
<span class="label label-danger">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a \`default color\` label 1`] = `
<span class="label label-secondary">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a \`info color\` label 1`] = `
<span class="label label-info">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a \`success color\` label 1`] = `
<span class="label label-success">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a \`warning color\` label 1`] = `
<span class="label label-warning">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a label with classes 1`] = `
<span class="label label-secondary my-custom-class">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a label with id 1`] = `
<span class="label label-secondary" id="myId">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render a label with label 1`] = `
<span class="label label-secondary">
  <span class="label-item label-item-expand">Foo</span>
</span>
`;

exports[`ClayLabel should render a label with link 1`] = `
<span class="label label-secondary">
  <span class="label-item label-item-expand">
    <a href="#1">Foo</a>
  </span>
</span>
`;

exports[`ClayLabel should render a label with removable 1`] = `
<span class="label label-secondary label-dismissible">
  <span class="label-item label-item-expand">Foo</span>
  <span class="label-item label-item-after">
    <button class="btn close btn-unstyled" aria-label="close" type="button">
      <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times" focusable="false">
        <title>times</title>
        <use xlink:href="icons.svg#times"></use>
      </svg>
    </button>
  </span>
</span>
`;

exports[`ClayLabel should render a large label 1`] = `
<span class="label label-secondary label-lg">
  <span class="label-item label-item-expand"></span>
</span>
`;

exports[`ClayLabel should render the default markup 1`] = `
<span class="label label-secondary">
  <span class="label-item label-item-expand"></span>
</span>
`;
